covenant/cache_params.go

```diff
@@ -0,0 +1,80 @@
+package covenant
+
+import (
+	"sync"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/babylonlabs-io/covenant-emulator/clientcontroller"
+	"github.com/babylonlabs-io/covenant-emulator/types"
+	"go.uber.org/zap"
+)
+
+type (
+	getParamByVersion func(version uint32) (*types.StakingParams, error)
+	ParamsGetter      interface {
+		Get(version uint32) (*types.StakingParams, error)
+	}
+	CacheVersionedParams struct {
+		sync.Mutex
+		paramsByVersion map[uint32]*types.StakingParams
+
+		getParamsByVersion getParamByVersion
```

Comment:
tbh given that `clientcontroller.ClientController` is an interface, I would use it directly in `CacheVersionedParams` i.e have field `cc clientcontroller.ClientController` in the struct.  This intermediate function `getParamByVersion` seems a bit unnecessary.